pkg/bundle/util.go

```diff
@@ -26,64 +27,42 @@ import (
-			// We ignore the LastTransitionTime as this is set by the controller
-			return existingCondition.Status == searchCondition.Status &&
-				existingCondition.Reason == searchCondition.Reason &&
-				existingCondition.Message == searchCondition.Message &&
-				existingCondition.ObservedGeneration == searchCondition.ObservedGeneration
-		}
+	existingCondition := meta.FindStatusCondition(existingConditions, searchCondition.Type)
+	if existingCondition == nil {
+		return false
 	}
-
-	return false
+	// Compare all fields except LastTransitionTime
+	// We ignore the LastTransitionTime as this is set by the controller
+	return existingCondition.Status == searchCondition.Status &&
+		existingCondition.Reason == searchCondition.Reason &&
+		existingCondition.Message == searchCondition.Message &&
+		existingCondition.ObservedGeneration == searchCondition.ObservedGeneration
 }
 
 // setBundleCondition updates the bundle with the given condition.
 // Will overwrite any existing condition of the same type.
 // LastTransitionTime will not be updated if an existing condition of the same
 // Type and Status already exists.
 func (b *bundle) setBundleCondition(
-	existingConditions []trustapi.BundleCondition,
-	patchConditions *[]trustapi.BundleCondition,
-	newCondition trustapi.BundleCondition,
-) trustapi.BundleCondition {
-	newCondition.LastTransitionTime = &metav1.Time{Time: b.clock.Now()}
-
-	// Reset the LastTransitionTime if the status hasn't changed
-	for _, cond := range existingConditions {
-		if cond.Type != newCondition.Type {
-			continue
-		}
-
-		// If this update doesn't contain a state transition, we don't update
-		// the conditions LastTransitionTime to Now()
-		if cond.Status == newCondition.Status {
-			newCondition.LastTransitionTime = cond.LastTransitionTime
+	existingConditions []metav1.Condition,
+	patchConditions *[]metav1.Condition,
+	newCondition metav1.Condition,
+) metav1.Condition {
+	existingCondition := meta.FindStatusCondition(existingConditions, newCondition.Type)
+	// If this update doesn't contain a state transition, we don't update
+	// the conditions LastTransitionTime to Now()
+	if existingCondition != nil && existingCondition.Status == newCondition.Status {
+		newCondition.LastTransitionTime = existingCondition.LastTransitionTime
+	} else {
+		if newCondition.LastTransitionTime.IsZero() {
+			newCondition.LastTransitionTime = metav1.Time{Time: b.clock.Now()}
 		}
 	}
 
-	// Search through existing conditions
-	for idx, cond := range *patchConditions {
-		// Skip unrelated conditions
-		if cond.Type != newCondition.Type {
-			continue
-		}
-
-		// Overwrite the existing condition
-		(*patchConditions)[idx] = newCondition
-
-		return newCondition
-	}
-
-	// If we've not found an existing condition of this type, we simply insert
-	// the new condition into the slice.
-	*patchConditions = append(*patchConditions, newCondition)
-
+	meta.SetStatusCondition(patchConditions, newCondition)
```

Comment:
`meta.SetStatusCondition` contains some code paths that we won't use (eg. setting LastTransitionTime).
I think that explicitly defining only the necessary logic here will make this function less error-prone.
The underlying reason is that `meta.SetStatusCondition` was created for CSA, not SSA.

Reply:
WDYT about removing the whole thing, and just append the new condition unconditionally? It's a bit counter-intuitive, at least to me, to account for overriding conditions when building up a SSA patch. To me, that's a code smell...

Reply:
I don't think we should silently allow you to create an invalid set of conditions.
We could throw an error if the condition is already set, but that code might be equally complex.

Reply:
That is actually a good idea! I will add the code to validate this, and I think it will be very simple.

Reply:
Ok, I have added the validation. Now with a panic, but I could change the signature to return an error.